### DISCUSSION TITLE
List and Count Method on Respository

### DISCUSSION THREAD

**Discussion by edimedia** [litestar-org/advanced-alchemy#205] [Q&A]
Hi! I want to know why the authors chose to make a count request to the DB instead of using `len(instances)`.

[https://github.com/litestar-org/advanced-alchemy/blob/24f7237ee024aa531ac54507cc66f5786245231f/advanced_alchemy/repository/_async.py#L1081](url)

            count = count_result.scalar_one()
            result = await self._execute(statement, loader_options_have_wildcards=loader_options_have_wildcard)
            instances: list[ModelT] = []
            for (instance,) in result:
                self._expunge(instance, auto_expunge=auto_expunge)
                instances.append(instance)
            return instances, count

instead of 

            result = await self._execute(statement, loader_options_have_wildcards=loader_options_have_wildcard)
            instances: list[ModelT] = []
            for (instance,) in result:
                self._expunge(instance, auto_expunge=auto_expunge)
                instances.append(instance)
            return instances, len(instances)          # <=========

I apologize if this question is not appropriate for this context.

**Comment by cofin** [comment]
@edimedia, the `list_and_count` method returns a tuple of the rows you selected and the "total rows" available.  This is typically used in places where you want to do pagination.  

Consider a case where you have passed in a `PaginationFilter`.  In this case, you want your `data` to be the limited result set, however, you need the total row count in most cases to determine your "current page".  In short we return `(<query results>, <count of results without any `limit` filters applied>)`

Depending on the database engine you are using, Advanced Alchemy will either add in the row count with a analytical window statement.  This lets us get the limited rows and the total count in a single query.  Otherwise, it issues 2 queries.  1 for the total count, and 1 for the data requested.

Hope this helps.

**Reply by edimedia** [reply]
Thanks, that's clear !
